### PR TITLE
graph(tests): fix for gcc 10.1 CTAD bug

### LINKED ISSUE #7316: Nightly Trilinos integration build failure: cuda/11.4.2+gcc/10.1.0, Cuda backend, `TestGraph.hpp:525:117: error: no matching function for call to 'Test::FetchValuesAndContribute`...
**Describe the bug**

Nightly Kokkos integration test builds with cuda/11.4.2+gcc/10.1.0 posting here https://trilinos-cdash.sandia.gov/build/1851535:  are failing to compile with output:

```
Standard Error 	

trilinos/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp: In member function 'virtual void Test::serial_graph_end_of_submit_control_flow_Test::TestBody()':
[workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:525](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L525):117: error: no matching function for call to 'Test::FetchValuesAndContribute<Kokkos::View<int*, Kokkos::Serial>, 5, 2>::FetchValuesAndContribute(view_t&, const size_t [2], std::integral_constant<long unsigned int, 5>&, const int&)'
  525 |   Kokkos::parallel_for(
      |                                                                                                                     ^
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:374](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L374):1: note: candidate: 'Test::FetchValuesAndContribute<ViewType, TargetIndex, NumIndices>::FetchValuesAndContribute(ViewType, Kokkos::Array<long unsigned int, NumIndices>, std::integral_constant<long unsigned int, _Nm>, typename ViewType::value_type) [with ViewType = Kokkos::View<int*, Kokkos::Serial>; long unsigned int TargetIndex = 5; long unsigned int NumIndices = 2; typename ViewType::value_type = int]'
  374 |   FetchValuesAndContribute(ViewType data_,
      | ^ ~~~~~~~~~~~~~~~~~~~~~~
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:375](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L375):1: note:   no known conversion for argument 2 from 'const size_t [2]' {aka 'const long unsigned int [2]'} to 'Kokkos::Array<long unsigned int, 2, void>'
  374 |   FetchValuesAndContribute(ViewType data_,
      |                                          ~
  375 |                            Kokkos::Array<size_t, NumIndices> indices_,
      | ^       
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:369](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L369):1: note: candidate: 'Test::FetchValuesAndContribute<ViewType, TargetIndex, NumIndices>::FetchValuesAndContribute(ViewType, std::integral_constant<long unsigned int, _Nm>, typename ViewType::value_type) [with ViewType = Kokkos::View<int*, Kokkos::Serial>; long unsigned int TargetIndex = 5; long unsigned int NumIndices = 2; typename ViewType::value_type = int]'
  369 |   FetchValuesAndContribute(ViewType data_,
      | ^ ~~~~~~~~~~~~~~~~~~~~~~
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:369](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L369):1: note:   candidate expects 3 arguments, 4 provided
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:361](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L361):8: note: candidate: 'Test::FetchValuesAndContribute<Kokkos::View<int*, Kokkos::Serial>, 5, 2>::FetchValuesAndContribute(const Test::FetchValuesAndContribute<Kokkos::View<int*, Kokkos::Serial>, 5, 2>&)'
  361 | struct FetchValuesAndContribute {
      |        ^~~~~~~~~~~~~~~~~~~~~~~~
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:361](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L361):8: note:   candidate expects 1 argument, 4 provided
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:361](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L361):8: note: candidate: 'Test::FetchValuesAndContribute<Kokkos::View<int*, Kokkos::Serial>, 5, 2>::FetchValuesAndContribute(Test::FetchValuesAndContribute<Kokkos::View<int*, Kokkos::Serial>, 5, 2>&&)'
[CTest: warning matched] [workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp:361](https://github.com/trilinos/Trilinos//blob/dbf8aaa628e8e64f33c3ec05065ff2567b8f041d/workspace/Trilinos_nightly_pipeline/Trilinos/kokkos/core/unit_test/TestGraph.hpp#L361):8: note:   candidate expects 1 argument, 4 provided
```

with targets Kokkos_CoreUnitTest_Serial1 and Kokkos_CoreUnitTest_Cuda1

This began with changes merged between 9/5/2024 and 9/6/2024

This test is launched through the Trilinos AT testing framework, I'll need to extract a reproducer and post

**Please include the following for a minimal reproducer**

1. Compilers (with versions)
2. Kokkos release or commit used (i.e., the sha1 number)
3. Platform, architecture and backend
4. CMake configure command
5. Output from CMake configure command
6. Minimum, complete code needed to reproduce the bug
7. Command line needed to reproduce the bug
8. `KokkosCore_config.h` header file (generated during the build)
9. Please provide any additional relevant error logs

### PR DESCRIPTION
**Pull request by romintomasetti**
This fixes #7316.

### CODE REVIEW

File: core/unit_test/TestGraph.hpp [hunk 1]
```diff
@@ -564,9 +564,20 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), end_of_submit_control_flow) {
...
+#if defined(KOKKOS_COMPILER_GNU) && (1010 == KOKKOS_COMPILER_GNU)
+  #define GCC_1010_CTAD(data, index, num) <decltype(data), index, num>
+#else
+  #define GCC_1010_CTAD(...)
+#endif
+
   Kokkos::parallel_for(
       policy_t(exec_3, 0, 1),
-      FetchValuesAndContribute(data, {index_D(), index_E()}, index_F, value_F));
+      FetchValuesAndContribute GCC_1010_CTAD(data, index_F, 2)(data, {index_D(), index_E()}, index_F, value_F));
```
Comment (dalg24) [inline core/unit_test/TestGraph.hpp:580]: Considering that this is only used in one place I would prefer
```suggestion
#if defined(KOKKOS_COMPILER_GNU) && (1010 == KOKKOS_COMPILER_GNU)
      // Workaround CTAD bug
      FetchValuesAndContribute<data, index_F, 2>(data, {index_D(), index_E()}, index_F, value_F));
#else
      FetchValuesAndContribute(data, {index_D(), index_E()}, index_F, value_F));
#endif
```

Comment (romintomasetti) [reply core/unit_test/TestGraph.hpp]: :checkered_flag:

File: core/unit_test/TestGraph.hpp [hunk 2]
```diff
@@ -566,8 +566,14 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), end_of_submit_control_flow) {
 
   Kokkos::parallel_for(
       policy_t(exec_3, 0, 1),
+#if defined(KOKKOS_COMPILER_GNU) && (1010 == KOKKOS_COMPILER_GNU)
+      // Workaround CTAD bug, see 7316.
+// clang-format off
+      FetchValuesAndContribute<data, index_F, 2>(data, {index_D(), index_E()}, index_F, value_F));
```
Comment (aprokop) [inline core/unit_test/TestGraph.hpp:572]: This seems off. Should it be
```suggestion
      FetchValuesAndContribute<decltype(data), index_F, 2>(data, {index_D(), index_E()}, index_F, value_F));
```

Comment (romintomasetti) [reply core/unit_test/TestGraph.hpp:572]: Yep, thanks @dalg24 for the buggy suggestion (https://github.com/kokkos/kokkos/pull/7323#discussion_r1755285396) :laughing:

Comment (dalg24) [reply core/unit_test/TestGraph.hpp]: Any time

Comment (romintomasetti) [reply core/unit_test/TestGraph.hpp]: Lesson learned: I should not apply blindly what you write :wink:

**Comment by romintomasetti** [comment]
I hate your clang format.

**Review by dalg24 — APPROVED** [review]

**Review by aprokop — APPROVED** [review]

File: core/unit_test/TestGraph.hpp [hunk 3]
```diff
@@ -564,10 +564,16 @@ TEST_F(TEST_CATEGORY_FIXTURE(graph), end_of_submit_control_flow) {
   //      their policy if the defaulted graph implementation is used.
   graph.submit(exec_3);
 
+  // clang-format off
   Kokkos::parallel_for(
       policy_t(exec_3, 0, 1),
+#if defined(KOKKOS_COMPILER_GNU) && (1010 == KOKKOS_COMPILER_GNU)
```
Comment (crtrott) [inline core/unit_test/TestGraph.hpp:570]: How sure are we it's only 10.1? (i.e. neither 10.0 nor 10.2 have the issue)?

**Review by crtrott — CHANGES_REQUESTED** [review]
Blocking for the answer on how specific the bug is to 10.1

File: core/unit_test/TestGraph.hpp (hunk 3 above)
Comment (romintomasetti) [reply core/unit_test/TestGraph.hpp:570]: Yes, it was only for `GCC 10.1` from my experiments.

**Review by crtrott — APPROVED** [review]
Romin confirmed to me on Slack - and if we get reports for another version we just revisit.

**Review by ndellingwood — APPROVED** [review]

File: core/unit_test/TestGraph.hpp (hunk 3 above)
Comment (dalg24) [reply core/unit_test/TestGraph.hpp:570]: We check in the compiler explorer.  Maybe Romin can share a link.

**Comment by dalg24** [comment]
Unrelated segfaults with NVHPC

**Pull request merged**